src/compiler/checker.ts

```diff
@@ -6543,6 +6549,17 @@ namespace ts {
             }
         }
 
+        function signatureHasTypePredicate(signature: Signature): boolean {
+            return signature.resolvedTypePredicate !== undefined;
+        }
+
+        function getTypePredicateOfSignature(signature: Signature): TypePredicate | undefined {

```

Comment:
Another approach we could consider is the one used to resolve Type Parameter constraints. Instead of using `undefined` to indicate no type predicate and an `unresolvedTypePredicate` sentinel to indicate deferred resolution, use `undefined` to indicate deferred resolution and a `noTypePredicate` sentinel. See `noConstraintType` as an example.